File: LC/python/Insertion Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if head == None or head.next == None:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        p = head
        
        while p.next:
            if p.next.val<p.val:
                pre = dummy
                while pre.next.val <p.next.val:
                    pre = pre.next
                
                tmp = p.next
                p.next = tmp.next
                tmp.next = pre.next
                pre.next = tmp
            else:
                p=p.next
        return dummy.next

File: LC/java/Insertion Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy,tmp,position;
        dummy = new ListNode(0);
        
        while(head != null){
            tmp = dummy;
            while(tmp.next!=null && head.val>tmp.next.val){
                tmp=tmp.next;
            }
            position = head.next;
            head.next = tmp.next;
            tmp.next = head;
            head = position;
        }
        
        return dummy.next;
    }
} 